Add group attribute tests for VFD SWMR

kyang2014 opened the pull request:
1. Add the group attribute tests to the vfd_swmr. The tests include add/delete/modify attributes(including variable length string) and different attribute attribute storage. Also tests the storage transition from compact to dense and vice versa while adding/deleting attributes.

2. If named pipes are used, the communication between the writer and the reader is enforced for every attribute/group operation.

kyang2014 committed:
1. Merge the group attribute test code to vfd_swmr_group_writer.c. Typical tests are fine. Need to clean up code and add comments/descriptions.
2. Add group attribute tests to testvfdswmr.sh.in. Only use typical cases for the standard and the quick runs for saving time.

kyang2014 committed:
Make the modifying attribute value obvious

kyang2014 committed:
Correct the debugging mismatch issue, add more comments.

kyang2014 committed:
Clean up code and add comments. More to do.

kyang2014 committed:
More comments and shorten the test time a bit

kyang2014 committed:
correct the code a bit, add comments to the shell script.

kyang2014 committed:
Merge remote-tracking branch 'upstream/feature/vfd_swmr' into feature/vfd_swmr

kyang2014 committed:
Use -q option for the group attribute test.

kyang2014 committed:
Forgot passing group property list when merging the code

kyang2014 committed:
Add the old-styled group creation attribute tests.

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -440,123 +2928,90 @@ main(int argc, char **argv)
+        s.max_lag   = config.max_lag;
+    }
+
+    /* For attribute test, force the named pipe to communicate in every step. */
```
For attribute testing, may I know why the synchronization has to be very step?

kyang2014 replied:
For each operation(either group or attribute),  I want the hand-shake between to the reader and the writer. I just want to make sure that there is no reader-go-ahead issue like the general group test .

vchoi-hdfgroup replied:
But is it possible for the user to choose "every step" or "every n steps"?

kyang2014 replied:
No, not for attribute tests when named pipe is used. Since in every step, a group is created. This is one operation.  As long as this exists, a possible reader ahead issue exists and a fake reader verification failure error may happen.  That's why Ray uses 5 ticks(instead of the default 3) for readers to wait in the default group test.  Users can choose every n steps for the default group test. This is the original test that has an attribute added for each group.  Users can use that option to add attributes for the group. However, if you think without using the named pipe, it benefits the users with all the attribute tests for uses to choose either every n steps or every 1 step. I can change the code. Without using the named pipe, there is no reason to force the attribute operations in every step.

vchoi-hdfgroup replied:
Thank you for the explanation.

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -43,26 +53,59 @@ typedef struct {
+        "              The value of `at_pattern` is one of the following:\n"
+        "              `compact`              - Attributes added in compact storage\n"
+        "              `dense`                - An attribute added in dense storage\n"
+        "              `compact-to-dense`     - Attributes added first in compact\n"
```
Suggest to change the name to `compact-add-to-dense` to be similar to `dense-del-to-compact`.  
Also move this down to be together with `dense-del-to-compact`.
This may help in grasping the testing patterns.

kyang2014 replied:
OK will do.

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -43,26 +53,59 @@ typedef struct {
+        "                                       deleted, the storage changed to\n"
+        "                                       compact\n"
+        "              `modify`               - An attribute added then modified\n"
+        "              `vstr`                 - A VL string attribute added\n"
```
Suggest to change to `add-vstr` to be similar to `remove-vstr` and `modify-vstr`.

kyang2014 replied:
OK

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -43,26 +53,59 @@ typedef struct {
+        , .np_verify = 0   }
+
 
+//TODO: add at_pattern description
```
`//TODO.....`  ----Will this be removed?

kyang2014 replied:
Yes, thanks. I think I removed it already.

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -171,30 +244,1538 @@ state_init(state_t *s, int argc, char **argv)
+#if 0
+    if(max_compact < min_dense) 
+        printf("Minimum number of attributes stored in dense storage should be less than maximum number of attributes stored in compact storage.\n");
+#endif
```
Will this `#if 0/#endif`  be removed?

kyang2014 replied:
Yes, will do.

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -171,30 +244,1538 @@ state_init(state_t *s, int argc, char **argv)
+    }
+
+    /* The following comments are left here in case in the future we want to
+     * use HDF5 function to verify the attribute storage */
```
I am inclined to do the verification of storage now instead of later.

kyang2014 replied:
This is a question I want to check how you did this. I don't find an obvious API and documentation how this can be done. In your code, I see the following block:

```
/* Get the object information */
    if(H5Oget_native_info(did, &ninfo, H5O_NATIVE_INFO_HDR|H5O_NATIVE_INFO_META_SIZE) < 0) {
        printf("H5Oget_native_info failed\n");
        TEST_ERROR;
    }

    if(!which) {
        dbgf(2, "Verifying no cont\n");
        return(ninfo.hdr.nchunks == 1);

    } else if(which < cut_point) {
        dbgf(2, "Verifying storage compact & cont\n");
        return(ninfo.meta_size.attr.index_size == 0 && 
               ninfo.meta_size.attr.heap_size == 0 &&
               ninfo.hdr.nchunks > 1);
    } else {
        dbgf(2, "Verifying storage dense & no cont\n");
        return(ninfo.meta_size.attr.index_size != 0 && 
               ninfo.meta_size.attr.heap_size != 0 &&
               ninfo.hdr.nchunks == 1);
    }

```
Is the```
 ninfo.meta_size.attr.index_size == 0 && 
               ninfo.meta_size.attr.heap_size == 0
```  for compact and the ```
ninfo.meta_size.attr.index_size != 0 && 
               ninfo.meta_size.attr.heap_size != 0
``` for dense?

vchoi-hdfgroup replied:
ninfo.meta_size.attr.index_size  and ninfo.meta_size.attr.heap_size are metadata size for data structures used to track dense storage.  They which will be zero for compact storage but nonzero for dense storage.
ninfo.hdr.nchunks is the # of chunks for the object header.  This will be 1 if there is just one chunk but > 1 if there are continuation chunks.

kyang2014 replied:
Thanks, that's what I think. I will use <ninfo.meta_size.attr.index_size and ninfo.meta_size.attr.heap_size>  to check the dense or the compact in the code. The <ninfo.hdr.nchunks> is irrelevant to the group attribute case. So I can ignore.

kyang2014 committed:
1. Add the attribute storage check
2. Change the test names and remove unnecessary debugging information
Tested at jelly.

vchoi-hdfgroup commented on test/vfd_swmr_group_writer.c:
```diff
@@ -112,216 +157,2736 @@ state_init(state_t *s, int argc, char **argv)
+    u= max_compact +1;
+
+
+    // delete a number of attributes so that the attribute storage just becomes dense.
```
Use /*  .,,,, */ instead of //

kyang2014 replied:
Fixed.

kyang2014 committed:
Use /* */ instead of // for comments.

raylu-hdf commented on test/vfd_swmr_group_writer.c:
```diff
@@ -112,216 +157,2736 @@ state_init(state_t *s, int argc, char **argv)
+    argv += optind;
+
+    /* space for attributes */
+    if ((s->one_by_one_sid = H5Screate_simple(1, &dims, &dims)) < 0) {
```
I don't see where this ID is being closed.

kyang2014 replied:
This is used by add_attr(). Around line 433, ` hid_t sid = s->one_by_one_sid;
`

raylu-hdf replied:
That's fine.  But that ID should be closed in the end.

raylu-hdf commented on test/vfd_swmr_group_writer.c:
```diff
@@ -112,216 +157,2736 @@ state_init(state_t *s, int argc, char **argv)
+                s->use_named_pipes = false;
+                break;
+            case 'A':
+                if (strcmp(optarg, "compact") == 0)
```
These string operations should have the prefix `HD`.  So do other string functions in this test.

kyang2014 replied:
will do.

CreationNationInovation reviewed (changes requested):
kyang2014:feature/vfd_swmr

raylu-hdf commented on test/vfd_swmr_group_writer.c:
```diff
@@ -112,216 +157,2736 @@ state_init(state_t *s, int argc, char **argv)
+    HDsprintf(attrname,adname_format,max_compact+which,0);
+    if (H5Adelete(obj_id,attrname) < 0) {
+            H5_FAILED(); AT();
+            printf("H5Adelete failed\n");
```
We have mixed ways of printing out error messages.  Sometimes `printf`, sometimes `dbgf`.  When we have time, we should clean them up.

kyang2014 replied:
dbgf is not for printing out error messages. It is just for debugging. Agree that we may make the dbgf consistent.

raylu-hdf commented on test/vfd_swmr_group_writer.c:
```diff
@@ -364,9 +2938,11 @@ main(int argc, char **argv)
 
     personality = strstr(s.progname, "vfd_swmr_group_");
```
Please use `HDstrstr` here.

kyang2014 replied:
Will do.

raylu-hdf commented on test/vfd_swmr_group_writer.c:
```diff
@@ -364,9 +2938,11 @@ main(int argc, char **argv)
 
-    if (personality != NULL && strcmp(personality, "vfd_swmr_group_writer") == 0)
+    if (personality != NULL &&
+        strcmp(personality, "vfd_swmr_group_writer") == 0)
```
Should be `HDstrcmp`

kyang2014 replied:
Will do.

kyang2014 committed:
Replace all string functions with HDF's corresponding string functions.

raylu-hdf reviewed (approved):
I approve it.  But I think `s->one_by_one_sid` should be closed in the end.

kyang2014 commented:
> I approve it. But I think `s->one_by_one_sid` should be closed in the end.

Yes, thanks a lot. Get it fixed.

kyang2014 committed:
Close the space id used for attributes.
